src/bgw/job_stat.c

```diff
@@ -0,0 +1,348 @@
+										int32 bgw_job_id)
+{
+	TupleDesc	desc = RelationGetDescr(rel);
+	Datum		values[Natts_bgw_job_stat];
+	bool		nulls[Natts_bgw_job_stat] = {false};
+	CatalogSecurityContext sec_ctx;
+
+	Interval	zero_ival = {.time = 0,};
+
+	values[Anum_bgw_job_stat_job_id - 1] = Int32GetDatum(bgw_job_id);
+	values[Anum_bgw_job_stat_last_start - 1] = TimestampGetDatum(GetCurrentTimestamp());
+	values[Anum_bgw_job_stat_last_finish - 1] = TimestampGetDatum(DT_NOBEGIN);
+	values[Anum_bgw_job_stat_next_start - 1] = TimestampGetDatum(DT_NOBEGIN);
+	values[Anum_bgw_job_stat_total_runs - 1] = Int64GetDatum(1);
+	values[Anum_bgw_job_stat_total_duration - 1] = IntervalPGetDatum(&zero_ival);
+	values[Anum_bgw_job_stat_total_success - 1] = Int64GetDatum(0);
+	values[Anum_bgw_job_stat_total_failures - 1] = Int64GetDatum(0);
+	values[Anum_bgw_job_stat_consecutive_failures - 1] = Int32GetDatum(0);
+
+	/* This is udone by any of the end marks */
+	values[Anum_bgw_job_stat_last_run_success - 1] = BoolGetDatum(false);
+	values[Anum_bgw_job_stat_total_crashes - 1] = Int64GetDatum(1);
+	values[Anum_bgw_job_stat_consecutive_crashes - 1] = Int32GetDatum(1);
+
+	catalog_become_owner(catalog_get(), &sec_ctx);
+	catalog_insert_values(rel, desc, values, nulls);
+	catalog_restore_user(&sec_ctx);
+
+	return true;
+}
+
+static bool
+bgw_job_stat_insert_mark_start(int32 bgw_job_id)
+{
+	Catalog    *catalog = catalog_get();
+	Relation	rel;
+	bool		result;
+
+	rel = heap_open(catalog->tables[BGW_JOB_STAT].id, RowExclusiveLock);
+	result = bgw_job_stat_insert_mark_start_relation(rel, bgw_job_id);
+	heap_close(rel, RowExclusiveLock);
+
+	return result;
+}
+
+
+void
+ts_bgw_job_stat_mark_start(int32 bgw_job_id)
+{
+	if (bgw_job_stat_scan_job_id(bgw_job_id, bgw_job_stat_tuple_mark_start, NULL, NULL, RowExclusiveLock) == 0)
+		bgw_job_stat_insert_mark_start(bgw_job_id);
+
+}
+
+void
+ts_bgw_job_stat_mark_end(BgwJob *job, bool success)
+{
+	JobResult	res =
+	{
+		.job = job,
+		.success = success,
+	};
+
+	if (bgw_job_stat_scan_job_id(job->fd.id, bgw_job_stat_tuple_mark_end, NULL, &res, RowExclusiveLock) == 0)
+		elog(ERROR, "unable to find job statistics for job %d", job->fd.id);
+
+}
+
+bool
+ts_bgw_job_stat_end_was_marked(BgwJobStat *jobstat)
+{
+	return !TIMESTAMP_IS_NOBEGIN(jobstat->fd.last_finish);
+}
+
+void
+ts_bgw_job_stat_set_next_start(BgwJob *job, TimestampTz next_start)
+{
+	/* Cannot use DT_NOBEGIN as that's the value used to indicate "not set" */
+	if (next_start == DT_NOBEGIN)
+		elog(ERROR, "cannot set next start to -infinity");
+
+	if (bgw_job_stat_scan_job_id(job->fd.id, bgw_job_stat_tuple_set_next_start, NULL, &next_start, RowExclusiveLock) == 0)
+		elog(ERROR, "unable to find job statistics for job %d", job->fd.id);
+
+}
+
+bool
+ts_bgw_job_stat_should_execute(BgwJobStat *jobstat, BgwJob *job)
+{
+	if (jobstat == NULL)
+		/* Never previously run */
+		return true;
+	return jobstat->fd.consecutive_crashes < MAX_CRASHES;
```

Comment:
I thought the verdict was to remove a MAX_CRASHES check and just keep increasing the backoff?

Reply:
Good point. Fixed